src/lxc/network.c

```diff
@@ -3438,29 +3452,39 @@ static int lxc_setup_netdev_in_child_namespaces(struct lxc_netdev *netdev)
 			return -1;
 		}
 
-		err = lxc_ipv6_gateway_add(netdev->ifindex, netdev->ipv6_gateway);
-		if (err) {
-			err = lxc_ipv6_dest_add(netdev->ifindex, netdev->ipv6_gateway, 128);
-			if (err) {
-				errno = -err;
-				SYSERROR("Failed to add ipv6 dest for network device \"%s\"",
+		/* Setup device route if ipv6_gateway_dev is enabled */
+		if (netdev->ipv6_gateway_dev) {
+			err = lxc_ipv6_gateway_add(netdev->ifindex, NULL);
+			if (err < 0) {
+				SYSERROR("Failed to setup ipv6 gateway to network device \"%s\"",
 				         ifname);
+				return minus_one_set_errno(-err);
 			}
-
+		} else {
 			err = lxc_ipv6_gateway_add(netdev->ifindex, netdev->ipv6_gateway);
 			if (err) {
-				errno = -err;
-				SYSERROR("Failed to setup ipv6 gateway for network device \"%s\"",
-				         ifname);
+				err = lxc_ipv6_dest_add(netdev->ifindex, netdev->ipv6_gateway, 128);
+				if (err) {
+					errno = -err;
+					SYSERROR("Failed to add ipv6 dest for network device \"%s\"",
+						ifname);
+				}
 
-				if (netdev->ipv6_gateway_auto) {
-					char buf[INET6_ADDRSTRLEN];
-					inet_ntop(AF_INET6, netdev->ipv6_gateway, buf, sizeof(buf));
-					ERROR("Tried to set autodetected ipv6 "
-					      "gateway for network device "
-					      "\"%s\"", buf);
+				err = lxc_ipv6_gateway_add(netdev->ifindex, netdev->ipv6_gateway);
+				if (err) {
+					errno = -err;
+					SYSERROR("Failed to setup ipv6 gateway for network device \"%s\"",
+						ifname);
+
+					if (netdev->ipv6_gateway_auto) {
+						char buf[INET6_ADDRSTRLEN];
+						inet_ntop(AF_INET6, netdev->ipv6_gateway, buf, sizeof(buf));
```

Comment:
Is it safe to ignore errors here?

Reply:
@brauner this isn't actually new code, its the existing gateway setup code just showing as new as I moved it into an `else` statement on line https://github.com/lxc/lxc/pull/2973/files#diff-49260cced2f6f9c5617962928313c73aR3463.

To answer your actual question, I think it is safe, because the gateway setting has failed and the container start will fail anyway on line 3486. It is just used to print out the detected gateway IP for more information in the errors.

However it is not ideal to not be checking the error from this function as presumably if it does fail the error message would contain an uninitialised char array, although thats very unlikely to happen.

More of concern for me is that there are several error messages that dont contain the problem IP above here, and only in this specific scenario does it actually log the gateway IP being used.

Further more on line https://github.com/lxc/lxc/blob/a2f9a6706d6363faa1fb6a091679799a1d41f374/src/lxc/network.c#L3466 if that fails it really shouldnt proceed any further as it has already tried to add the gateway and failed.

I can a look at re-structuring this code, would you prefer that done in a separate PR or as part of this? 

Thanks

Reply:
Feel free to restructure the code! But yes, please do it in a separate pr.